typescript/web/src/components/app-lifecycle-manager/welcome-modal/welcome-modal.tsx

```diff
@@ -1,185 +1,357 @@
-    // in the "loading app" state when a new service worker is waiting AND
-    // the welcome modal is open.
-    // This never happens except in nominal user flows, but still
     if (
-      typeof window !== "undefined" &&
-      "serviceWorker" in navigator &&
-      window.workbox !== undefined
+      !process.env.STORYBOOK &&
+      router?.isReady &&
+      ((!(browserWarning && tryDespiteBrowserWarning !== "true") &&
+        !isServiceWorkerActive &&
+        paramModalWelcome !== "closed") ||
+        paramModalWelcome === "open")
     ) {
-      const wb = window.workbox;
-      // // This next 3 lines were removed because they caused a reload of the page
-      // // when the user clicked "Start Labelling"
-      // wb.addEventListener("controlling", (/* event: any */) => {
-      //   window.location.reload();
-      // });
-      // Send a message to the waiting service worker, instructing it to activate.
-      wb.messageSkipWaiting();
+      performWelcomeWorkflow({
+        router,
+        setBrowserError,
+        isServiceWorkerActive,
+        setIsServiceWorkerActive,
+        client,
+        setParamModalWelcome,
+        setIsLoadingWorkerAndDemo,
+      });
     }
-  }, [setHasUserClickedStart, setParamModalWelcome]);
+  }, [tryDespiteBrowserWarning, router?.isReady]);
+
+  // welcome => undefined
+  const handleGetStarted = useCallback(() => {
+    setHasUserTriedApp("hasUserTriedApp", "true", {
+      path: "/",
+      httpOnly: false,
+    });
+
+    setParamModalWelcome(undefined, "replaceIn");
+
+    // First image of tutorial dataset
+    router.push(tutorialDatasetFirstImageUrl);
+  }, []);
+
+  // welcome => undefined
+  const handleSkip = useCallback(() => {
+    setHasUserTriedApp("hasUserTriedApp", "true", {
+      path: "/",
+      httpOnly: false,
+    });
+
+    setParamModalWelcome(undefined, "replaceIn");
+  }, []);
+
+  const shouldShowBrowserErrorModal = browserWarning && browserError != null;
+
+  const shouldShowBrowserWarningModal =
+    router?.isReady &&
+    browserError == null &&
+    browserWarning &&
+    tryDespiteBrowserWarning !== "true" &&
+    hasUserTriedApp !== "true";
+
+  const shouldShowLoadingModal =
+    router?.isReady && hasUserTriedApp !== "true" && isLoadingWorkerAndDemo;
+
+  const shouldShowWelcomeModal =
+    router?.isReady &&
+    (paramModalWelcome === "open" ||
+      (paramModalWelcome !== "closed" &&
+        hasUserTriedApp !== "true" &&
+        !isLoadingWorkerAndDemo));
 
   return (
     <Modal
-      isOpen={isOpen}
+      isOpen={
+        shouldShowBrowserErrorModal ||
+        (!shouldShowBrowserErrorModal && shouldShowBrowserWarningModal) ||
+        (!shouldShowBrowserErrorModal &&
+          !shouldShowBrowserWarningModal &&
+          shouldShowLoadingModal) ||
+        (!shouldShowBrowserErrorModal &&
+          !shouldShowBrowserWarningModal &&
+          !shouldShowLoadingModal &&
+          shouldShowWelcomeModal)
```

Comment:
I don't know if you made this on purpose to be explicit and some conditions are useless, or if the case where `shouldShowBrowserErrorModal` and `shouldShowBrowserWarningModal` are both true and we specifically want to close the modal here.

If that's the second option, you can ignore this message. 
Else: 

I guess you added this to do a kind of sanity check on whether or not we are in an incorrect state.
For example `!shouldShowBrowserErrorModal && !shouldShowBrowserWarningModal && shouldShowLoadingModal`.

IMO, I think it would be better to split the sanity check from the `isOpen` condition. 
I would do something like 
```typescript
if (shouldShowBrowserErrorModal && shouldShowBrowserWarningModal || allTheOtherIncorrectCombinaisons ) {
  throw new Error("Invalid State, this shouldn't happen");
}
```
And then you can safely do something like 
```typescript
const isOpen = shouldShowBrowserErrorModal | shouldShowBrowserWarningModal | shouldShowLoadingModal | shouldShowWelcomeModal;
```

This way, you could "properly" handle the incorrect case. With the current implementation, you handle this case by not opening the modal.

Reply:
If `shouldShowBrowserErrorModal && shouldShowBrowserWarningModal`, we do want to show the modal, and that's what is done here (the first clause of the disjunction is true, so the disjunction is true).

The reason why I added such a complicated condition is that: 

1. There is A LOT of cases (I counted at least 96 different cases in an excel spreadsheet !), (Cookie, error, browser warning, error on the page behind the modal, 404 behind the modal, user clicked some button or not, etc...) so better be explicit and not fuck around.
2. I dont want the modal to be displayed without having a Content
3. So the condition for `isOpen` is exactly (done in a mechanical way) the conditions that display stuff in its content. (the Immediately-invoked Function Expressions inside the modal content)
4. In this Immediately-invoked Function Expressions in the content, we use a function with early returns, so I just combine the conditions of each early return in a disjunction, and use that as `isOpen`